File: sdk/java-client/src/test/java/dev/latent/sdk/PublicationIdentityTest.java
package dev.latent.sdk;
import java.util.Optional;

final class PublicationIdentityTest {
    private PublicationIdentityTest() { }
    private static void check(boolean value) { if (!value) throw new AssertionError("publication model contract"); }
    static void run() {
        String component = "sha256:" + "a".repeat(64);
        var rows = new Models.PublicationIdentity[4];
        for (int i = 0; i < 4; ++i) rows[i] = new Models.PublicationIdentity(
            new Models.PublicationRef("publication:sha256:" + String.format("%064x", i), i < 2 ? "a" : "b"),
            component, "sha256:" + String.format("%064x", i % 2));
        check(rows[0].componentDigest().equals(rows[3].componentDigest()));
        check(rows[0].packageDigest().equals(rows[2].packageDigest()));
        check(!rows[0].packageDigest().equals(rows[1].packageDigest()));
        check(!rows[0].publication().equals(rows[2].publication()));
        var invalid = new Models.ReleaseSelector(Optional.of(""), Optional.of(new Models.PublicationRef("", "b")));
        check(invalid.componentDigest().orElseThrow().isEmpty() && invalid.publication().orElseThrow().id().isEmpty());
        var used = new Models.BudgetConsumption(-1L, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
        var legacy = new Models.InvocationReceipt("known", "revision", component, -1L, used);
        var current = new Models.InvocationReceipt("known", "revision", component, -1L, used,
            Optional.of(rows[1].publication().id()));
        check(legacy.publicationId().isEmpty() && current.publicationId().orElseThrow().equals(rows[1].publication().id()));
        check(current.releaseDigest().equals(component) && Long.toUnsignedString(current.routeGeneration()).equals("18446744073709551615"));
        check(Long.toUnsignedString(current.consumption().cpuFuel()).equals("18446744073709551615"));
    }
}

File: sdk/go/publication_identity_test.go
package latent_test

import (
	"fmt"
	latent "latent.dev/sdk/go"
	"testing"
)

func TestPublicationCoexistenceAndPresence(t *testing.T) {
	component := "sha256:" + fmt.Sprintf("%064x", 10)
	rows := make([]latent.PublicationIdentity, 4)
	ids := make(map[string]bool)
	for i := range rows {
		tenant := "a"
		if i >= 2 {
			tenant = "b"
		}
		rows[i] = latent.PublicationIdentity{Publication: latent.PublicationRef{
			ID: fmt.Sprintf("publication:sha256:%064x", i), Tenant: tenant},
			ComponentDigest: component, PackageDigest: fmt.Sprintf("sha256:%064x", i%2)}
		ids[rows[i].Publication.ID] = true
	}
	if len(ids) != 4 || rows[0].ComponentDigest != rows[3].ComponentDigest ||
		rows[0].PackageDigest != rows[2].PackageDigest || rows[0].PackageDigest == rows[1].PackageDigest {
		t.Fatal("publication, tenant, component and package identities must remain distinct")
	}
	invalid := latent.PublicationRef{ID: "", Tenant: "b"}
	if invalid.ID != "" || invalid.Tenant != "b" {
		t.Fatal("present invalid reference must not be normalized")
	}
	unresolved := latent.InvocationReceipt{ReleaseDigest: component, RouteGeneration: ^uint64(0),
		Consumption: latent.BudgetConsumption{CPUFuel: ^uint64(0)}}
	current := unresolved
	current.PublicationID = &rows[1].Publication.ID
	if unresolved.PublicationID != nil || *current.PublicationID != rows[1].Publication.ID ||
		current.ReleaseDigest != component || current.RouteGeneration != ^uint64(0) || current.Consumption.CPUFuel != ^uint64(0) {
		t.Fatal("receipt must preserve presence, component meaning and uint64 values")
	}
}
